python/dynamic programming/problem_0198_house_robber.py

```diff
@@ -0,0 +1,39 @@
+import unittest
+
+class Solution:
+    @staticmethod
+    def rob(nums):
+        if len(nums) <= 2:
+            return max(nums)
+        dp = [nums[0], nums[1]]
+        for i in range(2, len(nums)):
+            dp.append(nums[i] + max(dp[:-1]))
+        return max(dp[-1], dp[-2])
+
+"""
+Bottom up tabulation dp.
+
+Edge cases were basically what should we do when we encounter a house denoted with 0?
+Taking the max of the dp array excluding the most recent addition will assure that we don't hit the "0 house" while still excluding the immediately adjacent house. 
+
+Time complexity = O(n) tabulation dp. Only have to iterate through the array once
+Space = O(n) We have to make a completely separate dp array.
+"""
+
+class TestCase(unittest.TestCase):
+    def setUp(self):
+        self.solution = Solution()
+
+        self.testCase1 = [1,2,3,1]
```

Comment:
Move these variables into their respective testing functions— no need for them to be in the class scope

Reply:
Tests can be numerous, and their set-up can be repetitive. Luckily, we can factor out set-up code by implementing a method called [setUp()](https://docs.python.org/3/library/unittest.html#unittest.TestCase.setUp), which the testing framework will automatically call for every single test we run:

https://docs.python.org/3/library/unittest.html#module-unittest

That is correct, you only need to keep self.solution = Solution() in the setUp method, the rest should go into their respective test folders